File: php/src/Advertisement/Domain/Model/ValueObject/Email.php
<?php
declare(strict_types=1);

namespace Demo\App\Advertisement\Domain\Model\ValueObject;

use Chemem\Bingo\Functional\Functors\Monads\Either;
use Demo\App\Advertisement\Domain\Exceptions\InvalidEmailException;

final readonly class Email
{
    private function __construct(private string $value) {}

    public static function build(string $value): Either
    {
        if (!self::validate($value)) {
            return Either::left(InvalidEmailException::withEmail($value));
        }

        return Either::right(new self($value));
    }

    public function value(): string
    {
        return $this->value;
    }

    private static function validate(string $value): string|bool
    {
        return filter_var($value, FILTER_VALIDATE_EMAIL);
    }
}

File: kotlin/src/main/kotlin/advertisement/domain/model/value_object/Email.kt
package advertisement.domain.model.value_object

import advertisement.domain.errors.InvalidEmailFormatError
import arrow.core.Either
import arrow.core.left
import arrow.core.right

class Email private constructor (private val value: String) {

    fun value(): String = value
    companion object {
        fun build(value: String): Either<InvalidEmailFormatError, Email> {
            if (!validate(value)) {
                return InvalidEmailFormatError.withEmail(value).left()
            }
            return Email(value).right()
        }

        private fun validate(value: String): Boolean {
            val regex = "^[\\w-\\.]+@([\\w-]+\\.)+[\\w-]{2,4}\$"

            return value.matches(regex.toRegex())
        }
    }
}
